### PR TITLE
Hash table based on two variables

### PR DESCRIPTION
**Pull request by chengsijin0817**

### CODE REVIEW

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+import java.util.*;
+
+public class SolutionMappingsHashTableBasedOnTwoVars extends SolutionMappingsHashTable{
+    protected final Map<Node, List<Node>> mapNN = new HashMap<>();
+    protected final Map<Node, List<SolutionMapping>> mapNS = new HashMap<>();
+    protected final Var joinVar1;
+    protected final Var joinVar2;
+
+    public SolutionMappingsHashTableBasedOnTwoVars( final Var joinVar1, final Var joinVar2 ) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:17]:
```suggestion
    public SolutionMappingsHashTableBasedOnTwoVars( final Var joinVar1, final Var joinVar2 ) {
        assert joinVar1 != null;
        assert joinVar2 != null;
        assert ! joinVar1.equals(joinVar2);

```

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    @Override
+    public boolean add( final SolutionMapping e ) {
+        final Binding solMapBinding = e.asJenaBinding();
+        final Node n1 = solMapBinding.get(joinVar1);
+        final Node n2 = solMapBinding.get(joinVar2);
+
+        if (n1 == null || n2 == null){
+            throw new IllegalArgumentException();
+        }
+        else{
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:31]: No need to put this into an else-block. Removing the else-block will make the code a bit more readable.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            List<Node> varList = mapNN.get(n1);
+            if( varList == null){
+                varList = new ArrayList<>();
+                mapNN.put(n1, varList);
+            }
+            varList.add(n2);
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:37]: `n2` may already be contained in `varList`. In this case, there is no need to insert it twice.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+
+import java.util.*;
+
+public class SolutionMappingsHashTableBasedOnTwoVars extends SolutionMappingsHashTable{
+    protected final Map<Node, List<Node>> mapNN = new HashMap<>();
+    protected final Map<Node, List<SolutionMapping>> mapNS = new HashMap<>();
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:13]: I don't think that this is going to work. What would be the keys in `mapNS`? These keys should not be independent of the corresponding value for the other join variable. For instance, assume we first insert a solution mapping with var1=X and var2=Z, and after that we insert a solution mapping with var1=Y and var2=Z. Hence, both of these solution mappings have the same value for the second join variable. Then, when searching for join partners with var1=X and var2=Z, there should be an efficient way of retrieving only the relevant candidates with var1=X and var2=Z.

To fix this, the internal data structure here should be a nested map:
```
protected final Map<Node, Map<Node, List<SolutionMapping>>> map = new HashMap<>();
```
Then, the keys in the outer map (i.e., in `map`) are the RDF terms for the first join variable and the corresponding values would be further maps. The keys in each of these inner maps are the RDF terms for the second join variable, but only the ones that belong to the corresponding binding for the first variable.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
+            throws UnsupportedOperationException
+    {
+        final Binding solMapBinding = sm.asJenaBinding();
+        final Node n1 = solMapBinding.get(joinVar1);
+        final Node n2 = solMapBinding.get(joinVar2);
+        Iterator<SolutionMapping> matchingSolMaps;
+
+        if ( n1 == null || n2 == null ){
+            matchingSolMaps = iterator();
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:59]: This should be done only if `n1 == null`.

For the case that `n2 == null` but `n1 != null`, you can retrieve the inner map for `n1` and use an iterator over all elements in there, which reduces the number of candidates to be considered for the compatibility checks below.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+
+import java.util.*;
+
+public class SolutionMappingsHashTableBasedOnTwoVars extends SolutionMappingsHashTable{
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:11]: No. This class should be a subclass of `SolutionMappingsIndexBase`.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                joinPartner.add(matchSolM);
+            }
+        }
+        return joinPartner;
+    }
+
+    @Override
+    public Iterable<SolutionMapping> findSolutionMappings(
+            final Var var1, final Node value1,
+            final Var var2, final Node value2 )
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:79]: Where are the corresponding methods for one variable and for three variables?

**Review by hartig — CHANGES_REQUESTED** [review]

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java (hunk 4 above)
Comment (chengsijin0817) [reply src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java]: It makes sense. I have updated the data structure to a nested map.

File: src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java (hunk 7 above)
Comment (chengsijin0817) [reply src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java:159]: Added.

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpSymmetricHashJoin.java [hunk 8]
```diff
@@ -1,22 +1,24 @@
...
-import java.util.HashSet;
-import java.util.Set;
+import java.util.*;
 
 public class ExecOpSymmetricHashJoin implements BinaryExecutableOp{
 
-    protected final SolutionMappingsHashTable solMHashTableL;
-    protected final SolutionMappingsHashTable solMHashTableR;
+    protected final SolutionMappingsIndexBase solMHashTableL;
+    protected final SolutionMappingsIndexBase solMHashTableR;
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpSymmetricHashJoin.java:21]: Is it really necessary that these two are of type  SolutionMappingsIndexBase? Can't they just be of type SolutionMappingsIndex?

**Review by hartig — COMMENTED** [review]
Just one question.

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpSymmetricHashJoin.java (hunk 8 above)
Comment (chengsijin0817) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpSymmetricHashJoin.java:21]: Yes, we can declare these two as type SolutionMappingsIndex.
Currently, there is no other implementation besides SolutionMappingsIndexBase, and no distinct method is contained in SolutionMappingsIndexBase. Exactly, it is not necessary to declare them as SolutionMappingsIndexBase.

**Review by hartig — APPROVED** [review]
Looks all good now. Thanks.

**Pull request merged**